OB-6 Added a module scraping twitter time line posts

johnjs opened the pull request:
Hi!
This PR introduces a brand new twitter scraper! :fireworks:  Only the basic part has been implemented - the scraper fetches tweets from user's timeline using `statuses/user_timeline` endpoint.
I decided to follow the same approach to run both twitter and facebook scraper. Therefore few commits contain implementation of common/abstract runners, scrapers and feed stream.

When it comes to the schema of project's database, together with @asiakaczmar we agreed to use the same collection to store twitter and facebook scraping jobs. There are some small differences between twitter and facebook entities, but since mongodb lets us store objects with different structure in the same collection it's still fine and simplifies the codebase a lot.

[--> Trello ticket](https://trello.com/c/8qnqK8Fc/7-ob-7-implement-twitter-scraper)

johnjs committed:
OB-6 Implemented a base class for feed streams.

johnjs committed:
OB-6 Facebook feed stream extends abstract feed stream.

johnjs committed:
OB-6 Added a base class for scrapers.

johnjs committed:
OB-6 Facebook scraper extends the abstract scraper.

johnjs committed:
OB-6 Added Twit module as npm dependency.

johnjs committed:
OB-6 Added a module responsible for streaming twitter data

johnjs committed:
OB-6 Added twitter scraper

The scraper fetches last 10 tweets from user's timeline.

johnjs committed:
OB-6 Extracted a common `scraping job` entity to store both twitter and facebook jobs.

johnjs committed:
OB-6 Created a common runner for twitter and facebook jobs.

johnjs committed:
OB-6 Implemented module utils.

johnjs committed:
OB-6 Added proxyquire as npm dev dependency.

asiakaczmar commented on server/lib/abstract_feed_stream.js:
```diff
@@ -0,0 +1,23 @@
+   * specific services. Needs to be implemented in child classes.
+   **/
+  streamFeed() {
+    throw new Error('The "streamFeed" method must be implemented!');
```
Maybe you could move this string to some constant? We would avoid looking for all the places where it was used, if we want to change the message in the future.

johnjs replied:
Nice catch @asiakaczmar! :star: I've just implemented `NotImplementedError` class. Pushed in f712cbab909.

asiakaczmar commented on server/lib/facebook/facebook_feed_stream.js:
```diff
@@ -2,30 +2,23 @@
  **/
 
-import { PassThrough } from 'stream';
+import AbstractFeedStream from './../abstract_feed_stream';
```
You could remove the leading dot for better readability. Maybe it's also worth thinking of setting a NODE_PATH to avoid using relative paths?

johnjs replied:
@asiakaczmar, it's a quite [hot topic](https://news.ycombinator.com/item?id=7270436) :bomb: I've added a separate [ticket](https://trello.com/c/DfY7S7XQ/11-ob-11-get-rid-of-long-require-paths) to investigate and solve this problem. Thanks for pointing this out! :+1:

asiakaczmar commented on server/lib/twitter/twitter_feed_stream.js:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  _buildParametersForNextPageRequest(originalReqParams, lastDataResponse) {
+    const maxId = chain(lastDataResponse).pluck('id').min().value();
```
could you add a comment about why the min id in returned tweets should be used as max id for the next request?

johnjs replied:
Yeah, it may not be obvious when you look at it for the first time :smiley_cat: I put a piece of Twitter spec in the documentation of this method. Added in 0e5c541d. Thanks!

asiakaczmar commented on server/lib/scraping_job_schema.js:
```diff
@@ -1,23 +1,24 @@
 * @returns {Promise} resolving to the list of jobs
 **/
-facebookJobSchema.statics.findJobsToRun = function findJobsToRun() {
+scrapingJobSchema.statics.findJobsToRun = function findJobsToRun() {
```
Consider the following situation:
You have your polling interval (looking for jobs to run) set to one minute. If one of your jobs is running for longer then a minute, it will be pushed again to the queue, as the sheduleNextRun function is not called until the job is finished.
This could result in the same jobs being pushed to the queue over and over again.
There are several solutions. One would be to set a field on the object indicating that it was already sent to the queue, to avoid the additional runs before we update the nextRun field. Another way would be to update nextRun field immediately after queuing the job. In case of an error, you could always override the field to try again immediately, or wait few minutes and then try again.

johnjs replied:
That's a valid point @asiakaczmar :zap: I prefer the first solution. In the future I might think about implementing some kind of a timeout mechanism for forked processes, but the approach proposed in 6d31a88f6c should work fine.

johnjs committed:
OB-6 Implemented the NotImplementedError class.

johnjs committed:
OB-6 Added lacking documentation for twitter feed stream.

johnjs committed:
OB-6 Implemented the mechanism preventing the app from running multiple processes for single scraping job.

* defined the `is_running` property in scraping_job schema - once scraping of a given job
  is started its entry in db is marked as running. Running next scraping process for the job is blocked
  if the process pefroming the same action (started in the previous round of scheduling) is still active.
* defined timeouts for Twitter and Facebook API requests.

johnjs commented:
@asiakaczmar, I've addressed all your comments and pushed 3 extra commits to this PR. Could you have a look? :steam_locomotive: Thanks!

asiakaczmar commented:
:beer:  Very nice @johnjs !
